From 85598e5a888772c6f29d12506e1fc95dad104b12 Mon Sep 17 00:00:00 2001
From: Jian-Ming Wang <jian.ming.wang@mitaccomputing.com>
Date: Wed, 16 Jul 2025 03:56:12 +0000
Subject: [PATCH 2/2] Support MUXDEV and MUXCH so that configuration file for
 entity manager could use this information to create more modular design.
 Support MuxIdleMode to avoid confusion while sensor polling.

Upstream-Status: Inappropriate [No upstream maintenance]
---
 src/entity_manager/overlay.cpp | 119 +++++++++++++++++++++++++++++++--
 src/entity_manager/utils.cpp   |   3 +
 src/fru_device/fru_device.cpp  |  54 +++++++++++++++
 3 files changed, 170 insertions(+), 6 deletions(-)

diff --git a/src/entity_manager/overlay.cpp b/src/entity_manager/overlay.cpp
index 1a389fb..7413aa7 100644
--- a/src/entity_manager/overlay.cpp
+++ b/src/entity_manager/overlay.cpp
@@ -33,6 +33,7 @@
 #include <fstream>
 #include <iomanip>
 #include <iostream>
+#include <list>
 #include <regex>
 #include <string>
 
@@ -40,9 +41,54 @@ constexpr const char* outputDir = "/tmp/overlays";
 constexpr const char* templateChar = "$";
 constexpr const char* i2CDevsDir = "/sys/bus/i2c/devices";
 constexpr const char* muxSymlinkDir = "/dev/i2c-mux";
+constexpr const char* idleModeAsIs = "-1";
+constexpr const char* idleModeDisconnect = "-2";
 
 const std::regex illegalNameRegex("[^A-Za-z0-9_]");
 
+void setIdleMode(const std::string& muxName, size_t busIndex, size_t address,
+                 const std::string& idleMode)
+{
+    std::error_code ec;
+
+    std::ostringstream hexAddress;
+    hexAddress << std::hex << std::setfill('0') << std::setw(4) << address;
+
+    std::filesystem::path idlePath(i2CDevsDir);
+    idlePath /=
+        std::to_string(busIndex) + "-" + hexAddress.str() + "/idle_state";
+
+    std::string modeData =
+        (idleMode == "Disconnect") ? idleModeDisconnect : idleModeAsIs;
+
+    std::ofstream idleFile(idlePath);
+    if (!idleFile.good())
+    {
+        std::cerr << "Can't set idle mode in " << idlePath << " for " << muxName
+                  << "\n";
+    }
+    else
+    {
+        idleFile << modeData;
+    }
+}
+
+// Map for slot name to logical bus number
+using BusMap = boost::container::flat_map<std::string, uint64_t>;
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+static boost::container::flat_map<std::string, BusMap> muxChannelMap{};
+
+struct ExportInfo
+{
+    std::string type;
+    devices::ExportTemplate exportTemplate;
+    nlohmann::json configuration;
+    bool operator==(const ExportInfo& a) const
+    {
+        return configuration["Name"] == a.configuration["Name"];
+    }
+};
+
 // helper function to make json types into string
 std::string jsonToString(const nlohmann::json& in)
 {
@@ -217,7 +263,7 @@ static int buildDevice(
     return 0;
 }
 
-void exportDevice(const std::string& type,
+bool exportDevice(const std::string& type,
                   const devices::ExportTemplate& exportTemplate,
                   const nlohmann::json& configuration)
 {
@@ -230,6 +276,7 @@ void exportDevice(const std::string& type,
     std::optional<uint64_t> bus;
     std::optional<uint64_t> address;
     std::vector<std::string> channels;
+    std::string idleMode;
 
     for (auto keyPair = configuration.begin(); keyPair != configuration.end();
          keyPair++)
@@ -260,6 +307,34 @@ void exportDevice(const std::string& type,
         {
             channels = keyPair.value().get<std::vector<std::string>>();
         }
+        else if (keyPair.key() == "ChannelNames" && type.ends_with("Mux"))
+        {
+            channels = keyPair.value().get<std::vector<std::string>>();
+        }
+        else if (keyPair.key() == "MuxChannel")
+        {
+            auto muxChannel =
+                keyPair.value()
+                    .get<
+                        boost::container::flat_map<std::string, std::string>>();
+            if (muxChannelMap.contains(muxChannel["MuxName"]) &&
+                muxChannelMap[muxChannel["MuxName"]].contains(
+                    muxChannel["ChannelName"]))
+            {
+                bus = muxChannelMap[muxChannel["MuxName"]]
+                                   [muxChannel["ChannelName"]];
+                boost::replace_all(busPath, "$Bus", std::to_string(*bus));
+            }
+            else
+            {
+                return false;
+           }
+        }
+        else if (keyPair.key() == "MuxIdleMode")
+        {
+            idleMode = keyPair.value().get<std::string>();
+        }
+
         boost::replace_all(parameters, templateChar + keyPair.key(),
                            subsituteString);
         boost::replace_all(busPath, templateChar + keyPair.key(),
@@ -268,17 +343,26 @@ void exportDevice(const std::string& type,
 
     if (!bus || !address)
     {
-        createDevice(busPath, parameters, constructor);
-        return;
+        std::cout << "Create device at: " << busPath << "\n";
+        return (createDevice(busPath, parameters, constructor) >= 0);
     }
 
-    buildDevice(name, busPath, parameters, *bus, *address, constructor,
-                destructor, hasHWMonDir, std::move(channels));
+    if (buildDevice(name, busPath, parameters, *bus, *address, constructor,
+                        destructor, hasHWMonDir, std::move(channels)) < 0){
+        return false;
+    }
+    if (!idleMode.empty())
+    {
+            setIdleMode(name, static_cast<size_t>(*bus),
+                        static_cast<size_t>(*address), idleMode);
+    }
+    return true;
 }
 
 bool loadOverlays(const nlohmann::json& systemConfiguration)
 {
     std::filesystem::create_directory(outputDir);
+    std::list<ExportInfo> unfoundList;
     for (auto entity = systemConfiguration.begin();
          entity != systemConfiguration.end(); entity++)
     {
@@ -307,7 +391,11 @@ bool loadOverlays(const nlohmann::json& systemConfiguration)
             auto device = devices::exportTemplates.find(type.c_str());
             if (device != devices::exportTemplates.end())
             {
-                exportDevice(type, device->second, configuration);
+                if (!exportDevice(type, device->second, configuration))
+                {
+                    unfoundList.emplace_back(type, device->second,
+                                             configuration);
+                }
                 continue;
             }
 
@@ -320,5 +408,24 @@ bool loadOverlays(const nlohmann::json& systemConfiguration)
         }
     }
 
+    size_t unfoundListSize = unfoundList.size();
+    while (!unfoundList.empty())
+    {
+        for (auto& missingCfg : unfoundList)
+        {
+            if (exportDevice(missingCfg.type, missingCfg.exportTemplate,
+                             missingCfg.configuration))
+            {
+                unfoundList.remove(missingCfg);
+                break;
+            }
+        }
+        if (unfoundListSize == unfoundList.size())
+        {
+            std::cerr << "There is an incorrect config" << std::endl;
+            break;
+        }
+        unfoundListSize = unfoundList.size();
+    }
     return true;
 }
diff --git a/src/entity_manager/utils.cpp b/src/entity_manager/utils.cpp
index b169bce..4e9ea9f 100644
--- a/src/entity_manager/utils.cpp
+++ b/src/entity_manager/utils.cpp
@@ -8,6 +8,7 @@
 #include <boost/algorithm/string/predicate.hpp>
 #include <boost/algorithm/string/replace.hpp>
 #include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string/trim_all.hpp>
 #include <sdbusplus/bus/match.hpp>
 
 #include <fstream>
@@ -150,6 +151,7 @@ std::optional<std::string> templateCharReplace(
 
     boost::replace_all(*strPtr, std::string(templateChar) + "index",
                        std::to_string(index));
+    boost::trim_all(*strPtr);
     if (replaceStr)
     {
         boost::replace_all(*strPtr, *replaceStr, std::to_string(index));
@@ -184,6 +186,7 @@ std::optional<std::string> templateCharReplace(
         {
             std::string val = std::visit(VariantToStringVisitor(), propValue);
             boost::ireplace_all(*strPtr, templateName, val);
+            boost::trim_all(*strPtr);
             continue;
         }
 
diff --git a/src/fru_device/fru_device.cpp b/src/fru_device/fru_device.cpp
index c24d201..46c5d5d 100644
--- a/src/fru_device/fru_device.cpp
+++ b/src/fru_device/fru_device.cpp
@@ -179,6 +179,56 @@ static bool isMuxBus(size_t bus)
     return (!ec && isSymlink);
 }
 
+static std::optional<std::string> getMuxBusChannelName(size_t bus)
+{
+    if (!fs::exists("/dev/i2c-mux"))
+    {
+        return std::nullopt;
+    }
+
+    auto ec = std::error_code();
+    auto devName = "i2c-" + std::to_string(bus);
+    for (auto& mux : fs::recursive_directory_iterator("/dev/i2c-mux"))
+    {
+        auto channel = std::filesystem::read_symlink(mux, ec);
+        if (ec)
+        {
+            continue;
+        }
+
+        if (devName == channel.filename())
+        {
+            return mux.path().filename();
+        }
+    }
+    return std::nullopt;
+}
+
+static std::optional<std::string> getMuxBusDevName(size_t bus)
+{
+    if (!fs::exists("/dev/i2c-mux"))
+    {
+        return std::nullopt;
+    }
+
+    auto ec = std::error_code();
+    auto devName = "i2c-" + std::to_string(bus);
+    for (auto& mux : fs::recursive_directory_iterator("/dev/i2c-mux"))
+    {
+        auto channel = std::filesystem::read_symlink(mux, ec);
+        if (ec)
+        {
+            continue;
+        }
+
+        if (devName == channel.filename())
+        {
+            return mux.path().parent_path().filename();
+        }
+    }
+    return std::nullopt;
+}
+
 static void makeProbeInterface(size_t bus, size_t address,
                                sdbusplus::asio::object_server& objServer)
 {
@@ -833,6 +883,8 @@ void addFruObjectToDbus(
         objServer.add_interface(productName, "xyz.openbmc_project.FruDevice");
     dbusInterfaceMap[std::pair<size_t, size_t>(bus, address)] = iface;
 
+    auto muxChannel = getMuxBusChannelName(bus);
+    auto muxDev = getMuxBusDevName(bus);
     for (auto& property : formattedFRU)
     {
         std::regex_replace(property.second.begin(), property.second.begin(),
@@ -882,6 +934,8 @@ void addFruObjectToDbus(
     // baseboard will be 0, 0
     iface->register_property("BUS", bus);
     iface->register_property("ADDRESS", address);
+    iface->register_property("MUXCH", muxChannel.value_or(""));
+    iface->register_property("MUXDEV", muxDev.value_or(""));
 
     iface->initialize();
 }
-- 
2.45.2

